Hooks logging, sending logs to apiary-reporter, transaction.startedAt

kuba-kubula opened the pull request.

netmilk commented on src/sandbox-hooks-code.coffee:
```diff
@@ -12,23 +12,26 @@ sandboxHooksCode = (hooksCode, callback) ->
 
   sandbox = new Pitboss(wrappedCode)
-  sandbox.run {libraries: {"_Hooks": '../../../lib/hooks', "console", "console"}}, (err, result) ->
+  sandbox.run {libraries: {'_Hooks': '../../../lib/hooks', 'console'}}, (err, result) ->
```
What's the difference between `'` and `"` in JS?

kuba-kubula replied:
None, but in coffee-script, `"` are used for string variables interpolation.

netmilk commented on src/hooks-log-sandboxed.coffee:
```diff
@@ -0,0 +1,12 @@
+hooksLog = require './hooks-log'
```
Is it really necessary to have this in separate file?

kuba-kubula replied:
It is needed, because for _sandboxed context libraries_ - only this file is required as `_log` property and used along the code.

netmilk replied:
Brilliant! I apologise, I didn't realise it.

netmilk commented on src/transaction-runner.coffee:
```diff
@@ -82,8 +88,42 @@ class TransactionRunner
     @configuration.emitter.emit 'test error', error, test if error
 
+
+  sandboxedHookResultsHandler: (err, data, results = {}, callback) ->
```
:+1: for breaking it down to separate methods

netmilk commented on src/transaction-runner.coffee:
```diff
@@ -82,8 +88,42 @@ class TransactionRunner
+    return
+
+
+  sandboxedWrappedCode: (hookCode) ->
```
:+1:  ditto

netmilk commented on src/transaction-runner.coffee:
```diff
@@ -195,15 +215,16 @@ class TransactionRunner
 
-    parsedUrl = url.parse configuration['server']
+    # parse the server URL just once
+    @parsedUrl ?= url.parse configuration['server']
```
:+1:

netmilk commented on test/fixtures/hooks_log.coffee:
```diff
@@ -0,0 +1,8 @@
+{before, after, log} = require 'hooks'
```
File naming convention is: lowercase, dash separated imho

kuba-kubula replied:
this is a `testfile`, copying file naming conventions from other test files.

netmilk commented on test/fixtures/multifile/multifile_hooks.coffee:
```diff
@@ -1,10 +1,10 @@
-{after} = require 'hooks'
+{after, log} = require 'hooks'
```
I'm strictly against modifying existing fixtures and tests especially when this functionality should preserved. Please create separate fixture utilising the `hooks.log` and leave this `console.log` one for backward compatibility,

netmilk commented on test/fixtures/test2_all.js:
```diff
@@ -1,13 +1,13 @@
-var hooks;
+var hooks, log;
```
ditto

netmilk replied:
Why are you declaring `log` variable when you are not using later?

netmilk commented on test/fixtures/test_events.coffee:
```diff
@@ -1,5 +1,5 @@
+{afterAll,log} = require 'hooks'
 
 afterAll (done) ->
-  console.log "afterAll"
```
ditto, do not modify existing fixtures

netmilk commented on test/fixtures/test_hooks.coffee:
```diff
@@ -1,4 +1,4 @@
+{after,log} = require 'hooks'
 
 after "Machines > Machines collection > Get Machines", (transaction) ->
-  console.log "after"
```
ditto, please do not modify existing fixtures

netmilk commented on test/integration/cli-test.coffee:
```diff
@@ -291,6 +291,185 @@ describe "Command line interface", () ->
+      it 'should use toString when using log in hooks too', ->
+        assert.include (stdout + stderr), 'Error object!'
+
+      it 'should send result stats in PATCH request to Apiary with logs', ()->
```
Boom tschaca boom! :+1:

kuba-kubula commented:
@netmilk I've reverted the fixtures, restricted sandboxed hooks use of console (with a test for it) and fixed one more typo and one leftover (see last 2 commits).

Feel free to merge.

And btw, this should be `0.6` version, since it breaks usage of `console.log` in sandboxed hooks.

netmilk commented on src/transaction-runner.coffee:
```diff
@@ -117,32 +157,22 @@ class TransactionRunner
+        context:
+          "_data": data
+          stash: @hookStash
+          logs: []
```
This should be imho `_` prefixed as well. It's kind of a private variable, which we don't want to let users manipulate.

kuba-kubula commented:
@netmilk is it better now? `_logs` and deduplication. And one more test for transaction-runner.

netmilk commented on src/hooks-log.coffee:
```diff
@@ -0,0 +1,23 @@
+util = require 'util'
+
+hooksLog = (logs = [], logger, logVariant, content) ->
+  if arguments.length is 4 and logVariant in ['info', 'debug', 'warn', 'verbose', 'error', 'log']
```
Can you please articulate your motivation for this? I think that it's never used and it's only making the interface more confusing.

kuba-kubula replied:
It allows you to have logs more verbose, if you want to, according to the log-level you are using.
Because it uses the same `logger` as the rest of Dredd, it actually __is usable__ when used as
`hooks.log('verbose', 'Just starting validations')`

netmilk replied:
Ok, I get it, but where is this functionality used? Is it exposed to hooks users? I think that user facing `hook.log` should preserve same interface as native `console.log`.

kuba-kubula replied:
I can do it as if `hooks.log` would behave as calling `logger.info some, things, that, are, long` (as it works for winston - our logging platform).

But I would really appreciate to use just `hooks.log` as simply as possibly. So no, it is not the same as `console.log` since it obeys the used log-level.

netmilk commented:
I apologize, but I miss documentation update for this feature. May you please?

netmilk commented:
With this feature I mean `hooks.log` and `log`.

kuba-kubula commented:
@netmilk I might be missing some places when updating the documentation, but do we want the `hooks.log` function to be described with a separate example, or is __usage of `hooks.log`__ in other examples "enough"? If the usage is ok, can we merge this and bump to 0.6 please?

netmilk commented:
@kuba-kubula I think that it would be nice to have section called "Logging and Debugging". What do you think?

netmilk commented:
🍬
